framework/base/ComponentTrait.php

```diff
@@ -0,0 +1,595 @@
+    public function __get($name)
+    {
+        $getter = 'get' . $name;
+        if (method_exists($this, $getter)) {
+            // read property, e.g. getName()
+            return $this->$getter();
+        } else {
+            // behavior property
+            $this->ensureBehaviors();
+            foreach ($this->_behaviors as $behavior) {
+                if ($behavior->canGetProperty($name)) {
+                    return $behavior->$name;
+                }
+            }
+        }
+        if (method_exists($this, 'set' . $name)) {
+            throw new InvalidCallException('Getting write-only property: ' . get_class($this) . '::' . $name);
+        } else {
+            throw new UnknownPropertyException('Getting unknown property: ' . get_class($this) . '::' . $name);
+        }
+    }
+
+    /**
+     * Sets the value of a component property.
+     * This method will check in the following order and act accordingly:
+     *
+     *  - a property defined by a setter: set the property value
+     *  - an event in the format of "on xyz": attach the handler to the event "xyz"
+     *  - a behavior in the format of "as xyz": attach the behavior named as "xyz"
+     *  - a property of a behavior: set the behavior property value
+     *
+     * Do not call this method directly as it is a PHP magic method that
+     * will be implicitly called when executing `$component->property = $value;`.
+     * @param string $name the property name or the event name
+     * @param mixed $value the property value
+     * @throws UnknownPropertyException if the property is not defined
+     * @throws InvalidCallException if the property is read-only.
+     * @see __get()
+     */
+    public function __set($name, $value)
+    {
+        $setter = 'set' . $name;
+        if (method_exists($this, $setter)) {
+            // set property
+            $this->$setter($value);
+
+            return;
+        } elseif (strncmp($name, 'on ', 3) === 0) {
+            // on event: attach event handler
+            $this->on(trim(substr($name, 3)), $value);
+
+            return;
+        } elseif (strncmp($name, 'as ', 3) === 0) {
+            // as behavior: attach behavior
+            $name = trim(substr($name, 3));
+            $this->attachBehavior($name, $value instanceof Behavior ? $value : Yii::createObject($value));
+
+            return;
+        } else {
+            // behavior property
+            $this->ensureBehaviors();
+            foreach ($this->_behaviors as $behavior) {
+                if ($behavior->canSetProperty($name)) {
+                    $behavior->$name = $value;
+
+                    return;
+                }
+            }
+        }
+        if (method_exists($this, 'get' . $name)) {
+            throw new InvalidCallException('Setting read-only property: ' . get_class($this) . '::' . $name);
+        } else {
+            throw new UnknownPropertyException('Setting unknown property: ' . get_class($this) . '::' . $name);
```

Comment:
Overriding 2 methods is still a lot less work than implementing own behaviors/events logic. Current behavior is Yii convention which has some advantages (see https://github.com/yiisoft/yii2/issues/12021#issuecomment-236886911). If you you disagree with this, you can always change this pretty easy by implementing own traits with specified behavior.